x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -216,218 +231,383 @@ public WildcardFieldType clone() {
+            }
         }
 
-        private void addClause(String token, BooleanQuery.Builder bqBuilder) {
+        private void addClause(String token, BooleanQuery.Builder bqBuilder, Occur occur) {
             assert token.codePointCount(0, token.length()) <= NGRAM_SIZE;
             if (token.codePointCount(0, token.length()) == NGRAM_SIZE) {
                 TermQuery tq = new TermQuery(new Term(name(), token));
-                bqBuilder.add(new BooleanClause(tq, Occur.MUST));
+                bqBuilder.add(new BooleanClause(tq, occur));
             } else {
-                WildcardQuery wq = new WildcardQuery(new Term(name(), token + "*"));
-                wq.setRewriteMethod(MultiTermQuery.CONSTANT_SCORE_REWRITE);
-                bqBuilder.add(new BooleanClause(wq, Occur.MUST));
+                // Ignore tokens that are just string start or end markers
+                if (token.charAt(token.length() - 1) != TOKEN_START_OR_END_CHAR) {
+                    PrefixQuery wq = new PrefixQuery(new Term(name(), token));
+                    wq.setRewriteMethod(MultiTermQuery.CONSTANT_SCORE_REWRITE);
+                    bqBuilder.add(new BooleanClause(wq, occur));
+                }
             }
+        }
+
+        @Override
+        public Query fuzzyQuery(
+            Object value,
+            Fuzziness fuzziness,
+            int prefixLength,
+            int maxExpansions,
+            boolean transpositions,
+            QueryShardContext context
+        ) {
+            String searchTerm = BytesRefs.toString(value);
+            String lowerSearchTerm = toLowerCase(searchTerm);
+            TokenStream tokenizer = WILDCARD_ANALYZER.tokenStream(name(), lowerSearchTerm);
+            CharTermAttribute termAtt = tokenizer.addAttribute(CharTermAttribute.class);
+            ArrayList<String> tokens = new ArrayList<>();
+            String firstToken = null;
+            try {
+                tokenizer.reset();
+                int tokenNumber = 0;
+                while (tokenizer.incrementToken()) {
+                    if (tokenNumber == 0) {
+                        String token = termAtt.toString();
+                        if (firstToken == null) {
+                            firstToken = token;
+                        }
+                        tokens.add(token);
+                    }
+                    // Take every 3rd ngram so they are all disjoint. Our calculation for min_should_match
+                    // number relies on there being no overlaps
+                    tokenNumber++;
+                    if (tokenNumber == 3) {
+                        tokenNumber = 0;
+                    }
+                }
+                tokenizer.end();
+                tokenizer.close();
+
+                BooleanQuery.Builder bqBuilder = new BooleanQuery.Builder();
 
+                // Add any prefixLength as a MUST clause to the main BooleanQuery
```

Comment:
we could consider the entire prefix length instead of limiting to the first 3 characters ?